File: 21.template-method/templateMethod.java
// Clase abstracta
abstract class ReportGenerator {
    // Método plantilla
    public final void generateReport() {
        prepareData();
        createReport();
        saveReport();
    }

    protected abstract void prepareData();
    protected abstract void createReport();

    private void saveReport() {
        System.out.println("Guardando informe...");
    }
}

// Implementación concreta
class PDFReportGenerator extends ReportGenerator {
    @Override
    protected void prepareData() {
        System.out.println("Preparando datos para el informe PDF...");
    }

    @Override
    protected void createReport() {
        System.out.println("Creando informe PDF...");
    }
}

class ExcelReportGenerator extends ReportGenerator {
    @Override
    protected void prepareData() {
        System.out.println("Preparando datos para el informe Excel...");
    }

    @Override
    protected void createReport() {
        System.out.println("Creando informe Excel...");
    }
}

// Uso
public class templateMethod {
    public static void main(String[] args) {
        ReportGenerator pdfReport = new PDFReportGenerator();
        pdfReport.generateReport();

        ReportGenerator excelReport = new ExcelReportGenerator();
        excelReport.generateReport();
    }
}

File: 21.template-method/templateMethod.ts
// Clase base
abstract class ReportGenerator {
    // Método plantilla
    public generateReport(): void {
        this.prepareData();
        this.createReport();
        this.saveReport();
    }

    protected abstract prepareData(): void;
    protected abstract createReport(): void;

    private saveReport(): void {
        console.log("Guardando informe...");
    }
}

// Implementaciones concretas
class PDFReportGenerator extends ReportGenerator {
    protected prepareData(): void {
        console.log("Preparando datos para el informe PDF...");
    }

    protected createReport(): void {
        console.log("Creando informe PDF...");
    }
}

class ExcelReportGenerator extends ReportGenerator {
    protected prepareData(): void {
        console.log("Preparando datos para el informe Excel...");
    }

    protected createReport(): void {
        console.log("Creando informe Excel...");
    }
}

// Uso
const pdfReport = new PDFReportGenerator();
pdfReport.generateReport();

const excelReport = new ExcelReportGenerator();
excelReport.generateReport();